File: tests/py/include/extended_test_data.py
#!/usr/bin/env python3
"""
Extended test message data definitions (Python).
Hardcoded test messages for extended message ID and payload testing.

This module follows the C++ test pattern with a unified get_message(index) interface.
The get_message() function returns a message based on index, matching the C++ approach.
"""

import sys
import os
from typing import List, Union

# Add generated directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'generated', 'py'))

from struct_frame.generated.extended_test import (
    ExtendedTestExtendedIdMessage1,
    ExtendedTestExtendedIdMessage2,
    ExtendedTestExtendedIdMessage3,
    ExtendedTestExtendedIdMessage4,
    ExtendedTestExtendedIdMessage5,
    ExtendedTestExtendedIdMessage6,
    ExtendedTestExtendedIdMessage7,
    ExtendedTestExtendedIdMessage8,
    ExtendedTestExtendedIdMessage9,
    ExtendedTestExtendedIdMessage10,
    ExtendedTestLargePayloadMessage1,
    ExtendedTestLargePayloadMessage2,
    ExtendedTestExtendedVariableSingleArray,
    get_message_info,
)

# Type alias for message union (like C++ MessageVariant)
MessageType = Union[
    ExtendedTestExtendedIdMessage1,
    ExtendedTestExtendedIdMessage2,
    ExtendedTestExtendedIdMessage3,
    ExtendedTestExtendedIdMessage4,
    ExtendedTestExtendedIdMessage5,
    ExtendedTestExtendedIdMessage6,
    ExtendedTestExtendedIdMessage7,
    ExtendedTestExtendedIdMessage8,
    ExtendedTestExtendedIdMessage9,
    ExtendedTestExtendedIdMessage10,
    ExtendedTestLargePayloadMessage1,
    ExtendedTestLargePayloadMessage2,
    ExtendedTestExtendedVariableSingleArray,
]

# ============================================================================
# Message count and supported profiles
# ============================================================================

MESSAGE_COUNT = 17
PROFILES = "bulk, network"


# ============================================================================
# Helper functions to create messages (like C++ create_* functions)
# ============================================================================

def create_ext_id_1() -> ExtendedTestExtendedIdMessage1:
    return ExtendedTestExtendedIdMessage1(
        sequence_number=12345678,
        label=b"Test Label Extended 1",
        value=3.14159,
        enabled=True
    )


def create_ext_id_2() -> ExtendedTestExtendedIdMessage2:
    return ExtendedTestExtendedIdMessage2(
        sensor_id=-42,
        reading=2.718281828,
        status_code=50000,
        description=b"Extended ID test message 2"
    )


def create_ext_id_3() -> ExtendedTestExtendedIdMessage3:
    return ExtendedTestExtendedIdMessage3(
        timestamp=1704067200000000,
        temperature=-40,
        humidity=85,
        location=b"Sensor Room A"
    )


def create_ext_id_4() -> ExtendedTestExtendedIdMessage4:
    return ExtendedTestExtendedIdMessage4(
        event_id=999999,
        event_type=42,
        event_time=1704067200000,
        event_data=b"Event payload with extended message ID"
    )


def create_ext_id_5() -> ExtendedTestExtendedIdMessage5:
    return ExtendedTestExtendedIdMessage5(
        x_position=100.5,
        y_position=-200.25,
        z_position=50.125,
        frame_number=1000000
    )


def create_ext_id_6() -> ExtendedTestExtendedIdMessage6:
    return ExtendedTestExtendedIdMessage6(
        command_id=-12345,
        parameter1=1000,
        parameter2=2000,
        acknowledged=False,
        command_name=b"CALIBRATE_SENSOR"
    )


def create_ext_id_7() -> ExtendedTestExtendedIdMessage7:
    return ExtendedTestExtendedIdMessage7(
        counter=4294967295,
        average=123.456789,
        minimum=-999.99,
        maximum=999.99
    )


def create_ext_id_8() -> ExtendedTestExtendedIdMessage8:
    return ExtendedTestExtendedIdMessage8(
        level=255,
        offset=-32768,
        duration=86400000,
        tag=b"TEST123"
    )


def create_ext_id_9() -> ExtendedTestExtendedIdMessage9:
    return ExtendedTestExtendedIdMessage9(
        big_number=-9223372036854775807,
        big_unsigned=18446744073709551615,
        precision_value=1.7976931348623157e+308
    )


def create_ext_id_10() -> ExtendedTestExtendedIdMessage10:
    return ExtendedTestExtendedIdMessage10(
        small_value=256,
        short_text=b"Boundary Test",
        flag=True
    )


def create_large_1() -> ExtendedTestLargePayloadMessage1:
    sensor_readings = [float(i + 1) for i in range(64)]
    return ExtendedTestLargePayloadMessage1(
        sensor_readings=sensor_readings,
        reading_count=64,
        timestamp=1704067200000000,
        device_name=b"Large Sensor Array Device"
    )


def create_large_2() -> ExtendedTestLargePayloadMessage2:
    large_data = bytes([(i % 256) for i in range(280)])
    return ExtendedTestLargePayloadMessage2(
        large_data=large_data
    )


def create_ext_var_single_empty() -> ExtendedTestExtendedVariableSingleArray:
    """Create ExtendedVariableSingleArray with empty payload."""
    return ExtendedTestExtendedVariableSingleArray(
        timestamp=0x0000000000000001, telemetry_data=[], crc=0x00000001
    )


def create_ext_var_single_single() -> ExtendedTestExtendedVariableSingleArray:
    """Create ExtendedVariableSingleArray with single element."""
    return ExtendedTestExtendedVariableSingleArray(
        timestamp=0x0000000000000002, telemetry_data=[42], crc=0x00000002
    )


def create_ext_var_single_third() -> ExtendedTestExtendedVariableSingleArray:
    """Create ExtendedVariableSingleArray with 1/3 filled (83 elements for max_size=250)."""
    return ExtendedTestExtendedVariableSingleArray(
        timestamp=0x0000000000000003, telemetry_data=list(range(83)), crc=0x00000003
    )


def create_ext_var_single_almost() -> ExtendedTestExtendedVariableSingleArray:
    """Create ExtendedVariableSingleArray with 249 elements."""
    return ExtendedTestExtendedVariableSingleArray(
        timestamp=0x0000000000000004, telemetry_data=list(range(249)), crc=0x00000004
    )


def create_ext_var_single_full() -> ExtendedTestExtendedVariableSingleArray:
    """Create ExtendedVariableSingleArray with full 250 elements."""
    return ExtendedTestExtendedVariableSingleArray(
        timestamp=0x0000000000000005, telemetry_data=list(range(250)), crc=0x00000005
    )


# ============================================================================
# Unified get_message(index) function - matches C++ MessageProvider pattern
# ============================================================================

def get_message(index: int) -> MessageType:
    """
    Get message by index - unified interface matching C++ get_message().
    
    Message order: ExtendedId1-10, LargePayload1-2, ExtendedVariableSingleArray x5
    """
    if index == 0:
        return create_ext_id_1()
    elif index == 1:
        return create_ext_id_2()
    elif index == 2:
        return create_ext_id_3()
    elif index == 3:
        return create_ext_id_4()
    elif index == 4:
        return create_ext_id_5()
    elif index == 5:
        return create_ext_id_6()
    elif index == 6:
        return create_ext_id_7()
    elif index == 7:
        return create_ext_id_8()
    elif index == 8:
        return create_ext_id_9()
    elif index == 9:
        return create_ext_id_10()
    elif index == 10:
        return create_large_1()
    elif index == 11:
        return create_large_2()
    elif index == 12:
        return create_ext_var_single_empty()
    elif index == 13:
        return create_ext_var_single_single()
    elif index == 14:
        return create_ext_var_single_third()
    elif index == 15:
        return create_ext_var_single_almost()
    else:  # index == 16
        return create_ext_var_single_full()


# ============================================================================
# Message class lookup - maps message type to class for deserialization
# ============================================================================

MESSAGE_CLASSES = {
    ExtendedTestExtendedIdMessage1.MSG_ID: ExtendedTestExtendedIdMessage1,
    ExtendedTestExtendedIdMessage2.MSG_ID: ExtendedTestExtendedIdMessage2,
    ExtendedTestExtendedIdMessage3.MSG_ID: ExtendedTestExtendedIdMessage3,
    ExtendedTestExtendedIdMessage4.MSG_ID: ExtendedTestExtendedIdMessage4,
    ExtendedTestExtendedIdMessage5.MSG_ID: ExtendedTestExtendedIdMessage5,
    ExtendedTestExtendedIdMessage6.MSG_ID: ExtendedTestExtendedIdMessage6,
    ExtendedTestExtendedIdMessage7.MSG_ID: ExtendedTestExtendedIdMessage7,
    ExtendedTestExtendedIdMessage8.MSG_ID: ExtendedTestExtendedIdMessage8,
    ExtendedTestExtendedIdMessage9.MSG_ID: ExtendedTestExtendedIdMessage9,
    ExtendedTestExtendedIdMessage10.MSG_ID: ExtendedTestExtendedIdMessage10,
    ExtendedTestLargePayloadMessage1.MSG_ID: ExtendedTestLargePayloadMessage1,
    ExtendedTestLargePayloadMessage2.MSG_ID: ExtendedTestLargePayloadMessage2,
    ExtendedTestExtendedVariableSingleArray.MSG_ID: ExtendedTestExtendedVariableSingleArray,
}


# ============================================================================
# Encoder - simplified to use get_message(index) directly
# ============================================================================

class Encoder:
    """Encoder that uses get_message(index) for encoding."""
    
    def __init__(self):
        self._index = 0
    
    def write_message(self, writer, _msg_id: int) -> int:
        """Write message at current index to writer. Returns bytes written."""
        msg = get_message(self._index)
        self._index += 1
        return writer.write(msg)


# ============================================================================
# Validator - simplified to use get_message(index) directly
# ============================================================================

class Validator:
    """Validator that uses get_message(index) for validation."""
    
    def __init__(self):
        self._index = 0
    
    def validate_with_equals(self, frame_info) -> bool:
        """Validate decoded message using __eq__ operator."""
        msg_id = frame_info.msg_id
        msg_class = MESSAGE_CLASSES.get(msg_id)
        if not msg_class:
            return False
        
        expected = get_message(self._index)
        self._index += 1
        
        # Unpack both from packed bytes to ensure precision matches
        expected_unpacked = msg_class.deserialize(expected.serialize())
        decoded = msg_class.deserialize(frame_info)
        return decoded == expected_unpacked


# ============================================================================
# Test configuration - matches C++ TestHarness pattern
# ============================================================================

class Config:
    """Test configuration for extended messages."""
    MESSAGE_COUNT = MESSAGE_COUNT
    BUFFER_SIZE = 8192  # Larger for extended payloads
    FORMATS_HELP = PROFILES
    TEST_NAME = "Python Extended"
    
    @staticmethod
    def get_msg_id_order() -> List[int]:
        """Get message ID order array."""
        return [get_message(i).MSG_ID for i in range(MESSAGE_COUNT)]
    
    @staticmethod
    def create_encoder() -> Encoder:
        return Encoder()
    
    @staticmethod
    def create_validator() -> Validator:
        return Validator()
    
    @staticmethod
    def get_message_info(msg_id: int):
        return get_message_info(msg_id)
    
    @staticmethod
    def supports_format(format_name: str) -> bool:
        return format_name in ['bulk', 'network']


# Export config instance for convenience
extended_test_config = Config

File: tests/csharp/include/ExtendedTestData.cs
/**
 * Extended test message data definitions (C#).
 * Hardcoded test messages for extended message ID and payload testing.
 *
 * This module follows the same pattern as C, C++, TypeScript, and JavaScript
 * test data files.
 */

using System;
using StructFrame;
using StructFrame.ExtendedTest;

// Type aliases to match expected names
using ExtendedIdMessage1 = StructFrame.ExtendedTest.ExtendedTestExtendedIdMessage1;
using ExtendedIdMessage2 = StructFrame.ExtendedTest.ExtendedTestExtendedIdMessage2;
using ExtendedIdMessage3 = StructFrame.ExtendedTest.ExtendedTestExtendedIdMessage3;
using ExtendedIdMessage4 = StructFrame.ExtendedTest.ExtendedTestExtendedIdMessage4;
using ExtendedIdMessage5 = StructFrame.ExtendedTest.ExtendedTestExtendedIdMessage5;
using ExtendedIdMessage6 = StructFrame.ExtendedTest.ExtendedTestExtendedIdMessage6;
using ExtendedIdMessage7 = StructFrame.ExtendedTest.ExtendedTestExtendedIdMessage7;
using ExtendedIdMessage8 = StructFrame.ExtendedTest.ExtendedTestExtendedIdMessage8;
using ExtendedIdMessage9 = StructFrame.ExtendedTest.ExtendedTestExtendedIdMessage9;
using ExtendedIdMessage10 = StructFrame.ExtendedTest.ExtendedTestExtendedIdMessage10;
using LargePayloadMessage1 = StructFrame.ExtendedTest.ExtendedTestLargePayloadMessage1;
using LargePayloadMessage2 = StructFrame.ExtendedTest.ExtendedTestLargePayloadMessage2;

namespace StructFrameTests
{
    /// <summary>
    /// Wrapper for extended test messages that provides a common interface
    /// </summary>
    public class ExtendedTestMessage
    {
        public int MsgId { get; }
        public int MaxSize { get; }
        private readonly Func<byte[]> _packFunc;

        public ExtendedTestMessage(int msgId, int maxSize, Func<byte[]> packFunc)
        {
            MsgId = msgId;
            MaxSize = maxSize;
            _packFunc = packFunc;
        }

        public byte[] Pack() => _packFunc();
    }

    public static class ExtendedTestData
    {
        // ============================================================================
        // Message count
        // ============================================================================

        public const int MESSAGE_COUNT = 12;

        public static int GetExtendedTestMessageCount()
        {
            return MESSAGE_COUNT;
        }

        // ============================================================================
        // Message info lookup (unified size and magic numbers)
        // ============================================================================

        public static MessageInfo? GetMessageInfo(int msgId)
        {
            // Combined message ID (pkg_id << 8 | msg_id)
            switch (msgId)
            {
                case ExtendedIdMessage1.MsgId: return new MessageInfo(ExtendedIdMessage1.MaxSize, ExtendedIdMessage1.Magic1, ExtendedIdMessage1.Magic2);
                case ExtendedIdMessage2.MsgId: return new MessageInfo(ExtendedIdMessage2.MaxSize, ExtendedIdMessage2.Magic1, ExtendedIdMessage2.Magic2);
                case ExtendedIdMessage3.MsgId: return new MessageInfo(ExtendedIdMessage3.MaxSize, ExtendedIdMessage3.Magic1, ExtendedIdMessage3.Magic2);
                case ExtendedIdMessage4.MsgId: return new MessageInfo(ExtendedIdMessage4.MaxSize, ExtendedIdMessage4.Magic1, ExtendedIdMessage4.Magic2);
                case ExtendedIdMessage5.MsgId: return new MessageInfo(ExtendedIdMessage5.MaxSize, ExtendedIdMessage5.Magic1, ExtendedIdMessage5.Magic2);
                case ExtendedIdMessage6.MsgId: return new MessageInfo(ExtendedIdMessage6.MaxSize, ExtendedIdMessage6.Magic1, ExtendedIdMessage6.Magic2);
                case ExtendedIdMessage7.MsgId: return new MessageInfo(ExtendedIdMessage7.MaxSize, ExtendedIdMessage7.Magic1, ExtendedIdMessage7.Magic2);
                case ExtendedIdMessage8.MsgId: return new MessageInfo(ExtendedIdMessage8.MaxSize, ExtendedIdMessage8.Magic1, ExtendedIdMessage8.Magic2);
                case ExtendedIdMessage9.MsgId: return new MessageInfo(ExtendedIdMessage9.MaxSize, ExtendedIdMessage9.Magic1, ExtendedIdMessage9.Magic2);
                case ExtendedIdMessage10.MsgId: return new MessageInfo(ExtendedIdMessage10.MaxSize, ExtendedIdMessage10.Magic1, ExtendedIdMessage10.Magic2);
                case LargePayloadMessage1.MsgId: return new MessageInfo(LargePayloadMessage1.MaxSize, LargePayloadMessage1.Magic1, LargePayloadMessage1.Magic2);
                case LargePayloadMessage2.MsgId: return new MessageInfo(LargePayloadMessage2.MaxSize, LargePayloadMessage2.Magic1, LargePayloadMessage2.Magic2);
                default: return null;
            }
        }

        // ============================================================================
        // Message creation
        // ============================================================================

        private static byte[] CreateLabelBytes(string text, int size)
        {
            var bytes = new byte[size];
            var textBytes = System.Text.Encoding.UTF8.GetBytes(text);
            Array.Copy(textBytes, bytes, Math.Min(textBytes.Length, size));
            return bytes;
        }

        public static (ExtendedTestMessage message, string typeName) GetExtendedTestMessage(int index)
        {
            switch (index)
            {
                // Message 0: ExtendedIdMessage1 (matches C++ create_ext_id_1)
                case 0:
                {
                    var msg = new ExtendedIdMessage1
                    {
                        SequenceNumber = 12345678,
                        Label = CreateLabelBytes("Test Label Extended 1", 32),
                        Value = 3.14159f,
                        Enabled = true
                    };
                    return (new ExtendedTestMessage(ExtendedIdMessage1.MsgId, ExtendedIdMessage1.MaxSize, msg.Pack), "ExtendedIdMessage1");
                }

                // Message 1: ExtendedIdMessage2 (matches C++ create_ext_id_2)
                case 1:
                {
                    var msg = new ExtendedIdMessage2
                    {
                        SensorId = -42,
                        Reading = 2.718281828,
                        StatusCode = 50000,
                        DescriptionLength = 26,
                        DescriptionData = CreateLabelBytes("Extended ID test message 2", 64)
                    };
                    return (new ExtendedTestMessage(ExtendedIdMessage2.MsgId, ExtendedIdMessage2.MaxSize, msg.Pack), "ExtendedIdMessage2");
                }

                // Message 2: ExtendedIdMessage3 (matches C++ create_ext_id_3)
                case 2:
                {
                    var msg = new ExtendedIdMessage3
                    {
                        Timestamp = 1704067200000000,
                        Temperature = -40,
                        Humidity = 85,
                        Location = CreateLabelBytes("Sensor Room A", 16)
                    };
                    return (new ExtendedTestMessage(ExtendedIdMessage3.MsgId, ExtendedIdMessage3.MaxSize, msg.Pack), "ExtendedIdMessage3");
                }

                // Message 3: ExtendedIdMessage4 (matches C++ create_ext_id_4)
                case 3:
                {
                    var msg = new ExtendedIdMessage4
                    {
                        EventId = 999999,
                        EventType = 42,
                        EventTime = 1704067200000,
                        EventDataLength = 38,
                        EventDataData = CreateLabelBytes("Event payload with extended message ID", 128)
                    };
                    return (new ExtendedTestMessage(ExtendedIdMessage4.MsgId, ExtendedIdMessage4.MaxSize, msg.Pack), "ExtendedIdMessage4");
                }

                // Message 4: ExtendedIdMessage5 (matches C++ create_ext_id_5)
                case 4:
                {
                    var msg = new ExtendedIdMessage5
                    {
                        XPosition = 100.5f,
                        YPosition = -200.25f,
                        ZPosition = 50.125f,
                        FrameNumber = 1000000
                    };
                    return (new ExtendedTestMessage(ExtendedIdMessage5.MsgId, ExtendedIdMessage5.MaxSize, msg.Pack), "ExtendedIdMessage5");
                }

                // Message 5: ExtendedIdMessage6 (matches C++ create_ext_id_6)
                case 5:
                {
                    var msg = new ExtendedIdMessage6
                    {
                        CommandId = -12345,
                        Parameter1 = 1000,
                        Parameter2 = 2000,
                        Acknowledged = false,
                        CommandName = CreateLabelBytes("CALIBRATE_SENSOR", 24)
                    };
                    return (new ExtendedTestMessage(ExtendedIdMessage6.MsgId, ExtendedIdMessage6.MaxSize, msg.Pack), "ExtendedIdMessage6");
                }

                // Message 6: ExtendedIdMessage7 (matches C++ create_ext_id_7)
                case 6:
                {
                    var msg = new ExtendedIdMessage7
                    {
                        Counter = 4294967295,
                        Average = 123.456789,
                        Minimum = -999.99f,
                        Maximum = 999.99f
                    };
                    return (new ExtendedTestMessage(ExtendedIdMessage7.MsgId, ExtendedIdMessage7.MaxSize, msg.Pack), "ExtendedIdMessage7");
                }

                // Message 7: ExtendedIdMessage8 (matches C++ create_ext_id_8)
                case 7:
                {
                    var msg = new ExtendedIdMessage8
                    {
                        Level = 255,
                        Offset = -32768,
                        Duration = 86400000,
                        Tag = CreateLabelBytes("TEST123", 8)
                    };
                    return (new ExtendedTestMessage(ExtendedIdMessage8.MsgId, ExtendedIdMessage8.MaxSize, msg.Pack), "ExtendedIdMessage8");
                }

                // Message 8: ExtendedIdMessage9 (matches C++ create_ext_id_9)
                case 8:
                {
                    var msg = new ExtendedIdMessage9
                    {
                        BigNumber = -9223372036854775807,
                        BigUnsigned = 18446744073709551615,
                        PrecisionValue = 1.7976931348623157e+308
                    };
                    return (new ExtendedTestMessage(ExtendedIdMessage9.MsgId, ExtendedIdMessage9.MaxSize, msg.Pack), "ExtendedIdMessage9");
                }

                // Message 9: ExtendedIdMessage10 (matches C++ create_ext_id_10)
                case 9:
                {
                    var msg = new ExtendedIdMessage10
                    {
                        SmallValue = 256,
                        ShortText = CreateLabelBytes("Boundary Test", 16),
                        Flag = true
                    };
                    return (new ExtendedTestMessage(ExtendedIdMessage10.MsgId, ExtendedIdMessage10.MaxSize, msg.Pack), "ExtendedIdMessage10");
                }

                // Message 10: LargePayloadMessage1 (matches C++ create_large_1)
                case 10:
                {
                    var sensorReadings = new float[64];
                    for (int i = 0; i < 64; i++)
                    {
                        sensorReadings[i] = (float)(i + 1);
                    }
                    var msg = new LargePayloadMessage1
                    {
                        SensorReadings = sensorReadings,
                        ReadingCount = 64,
                        Timestamp = 1704067200000000,
                        DeviceName = CreateLabelBytes("Large Sensor Array Device", 32)
                    };
                    return (new ExtendedTestMessage(LargePayloadMessage1.MsgId, LargePayloadMessage1.MaxSize, msg.Pack), "LargePayloadMessage1");
                }

                // Message 11: LargePayloadMessage2 (matches C++ create_large_2)
                case 11:
                {
                    var largeData = new byte[280];
                    for (int i = 0; i < 256; i++)
                    {
                        largeData[i] = (byte)i;
                    }
                    for (int i = 256; i < 280; i++)
                    {
                        largeData[i] = (byte)(i - 256);
                    }
                    var msg = new LargePayloadMessage2
                    {
                        LargeData = largeData
                    };
                    return (new ExtendedTestMessage(LargePayloadMessage2.MsgId, LargePayloadMessage2.MaxSize, msg.Pack), "LargePayloadMessage2");
                }

                default:
                    return (null, null);
            }
        }

        // ============================================================================
        // Test configuration
        // ============================================================================

        public static class Config
        {
            public const int MESSAGE_COUNT = 12;
            public const int BUFFER_SIZE = 8192;
            public const string FORMATS_HELP = "profile_bulk, profile_network";
            public const string TEST_NAME = "extended";

            public static bool SupportsFormat(string formatName)
            {
                return formatName == "profile_bulk" || formatName == "profile_network";
            }
        }
    }
}
